Added window id parameter in _tabs funciton used it in delete_tabs function

mohite-abhi opened the pull request:
As delete_tab was assuming that completion column contains window ID, it was showing
exception in case of tab-focus, as it doesn't have the window ID in completion column.
So instead a new parameter named current_win_id is used in _tabs which is also passed
in all uses of the _tabs function. This PR Closes #6967.

<!-- Thanks for submitting a pull request! Please pick a descriptive title (not just "issue 12345"). If there is an open issue associated to your PR, please add a line like "Closes #12345" somewhere in the PR description (outside of this comment) -->

rcorre commented on qutebrowser/completion/models/miscmodels.py:
```diff
@@ -103,17 +103,21 @@ def session(*, info=None):
-        win_id, tab_index = data[0].split('/')
+        win_id = cur_win_id
+        # data[0] can be 'tabInd' or 'winID/tabInd'
+        tab_index = data[0].split('/')[-1]
```
I think we still need to get a window ID from `data`, and only fall back to `cur_win_id` if the former is missing. Otherwise we might delete the wrong tab if there are multiple windows.

rcorre reviewed (changes requested):
Could you add a test to https://github.com/mohite-abhi/qutebrowser/blob/d89d5853d393390b3a62c33084bf8150858a2fee/tests/unit/completion/test_models.py#L842-L843 that covers the single-window case?

Other than that and the one comment, it looks good. Thanks!

mohite-abhi commented:
I have added the asked test. I have also done some other changes as discussed [here](https://github.com/qutebrowser/qutebrowser/issues/6967#issuecomment-1020029076) (solution 1) and modified some tests to account for those changes. I hope I have done it correctly.

The-Compiler commented on qutebrowser/completion/models/miscmodels.py:
```diff
@@ -103,20 +103,28 @@ def session(*, info=None):
     def delete_tab(data):
         """Close the selected tab."""
-        win_id, tab_index = data[0].split('/')
+        unpacked_data = data[0].split('/')
+
+        if len(unpacked_data) == 1:
+            win_id = cur_win_id
+        else:
+            win_id = unpacked_data[0]
```
I think this would be a bit clearer as something like:

```suggestion
        if cur_win_id is None: 
            win_id, tab_index = data[0].split('/')
        else:
            win_id = current_win_id
            tab_index = data[0]
 ```

or so, with the default value changed accordingly. That way, accidental misuse of the function is prevented too.

The-Compiler reviewed (changes requested):
I don't really feel comfortable with the change to the format of `data` - it seems very odd to use `-1` as a special key, and I suspect it will need some closer look at other completion functions using deletion functionality too.

I would rather keep it separate from the bugfix in the original issue, so I would suggest that you revert the related changes as part of this PR and we look at them again as a separate PR once this one is in. What do you think?

mohite-abhi commented:
I think the crash issue should be fixed now which was happening due to python exception. Still there is some wierd behaviour on deletion of multiple tabs through tab focus completion, but yes we can look into it in another PR. 

Off topic : This is my first PR ever on an open source project, so i would like to thankyou @The-Compiler and @rcorre for your support. I like qutebrowser very much and it's the primary browser I use and hope to contribute as much as I can to the project :smile: .

The-Compiler commented:
Great work, @mohite-abhi! Thanks again and congrats on your first contribution! :tada: 

Feel free to open a separate PR for the other changes you proposed, then we can look at those in more detail. Let me note that I might not be able to look at them in the next 3 weeks or so, though.

mohite-abhi commented:
Thanks very much. Indeed I will try to make those changes
<https://github.com/qutebrowser/qutebrowser/issues/6967#issuecomment-1020029076>
.

On Sat, Jan 29, 2022 at 4:09 PM Florian Bruhin ***@***.***>
wrote:

> Great work, @mohite-abhi <https://github.com/mohite-abhi>! Thanks again
> and congrats on your first contribution! 🎉
>
> Feel free to open a separate PR for the other changes you proposed, then
> we can look at those in more detail. Let me note that I might not be able
> to look at them in the next 3 weeks or so, though.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/qutebrowser/qutebrowser/pull/6972#issuecomment-1024886142>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AJQCXEU3X63G72KAX5JXJYDUYO7XPANCNFSM5MTL5NSA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>
